lib/bolt/outputter.rb

```diff
@@ -21,6 +21,60 @@ def initialize(color, verbose, trace, stream = $stdout)
       @trace = trace
       @stream = stream
     end
+
+    def indent(indent, string)
+      indent = ' ' * indent
+      string.gsub(/^/, indent.to_s)
+    end
+
+    def print_message_event(event)
+      print_message(stringify(event[:message]))
+    end
+
+    def print_message
+      raise NotImplementedError, "print_message() must be implemented by the outputter class"
+    end
+
+    def stringify(message)
+      formatted = format_message(message)
+      if formatted.is_a?(Hash) || formatted.is_a?(Array)
+        ::JSON.pretty_generate(formatted)
+      else
+        formatted
+      end
+    end
+
+    def format_message(message)
+      case message
+      when Array
+        message.map { |item| format_message(item) }
+      when Bolt::ApplyResult
+        format_apply_result(message)
+      when Bolt::Result, Bolt::ResultSet
+        # This is equivalent to to_s, but formattable
+        message.to_data
+      when Bolt::RunFailure
+        formatted_resultset = message.result_set.to_data
+        message.to_h.merge('result_set' => formatted_resultset)
+      when Hash
+        message.each_with_object({}) do |(k, v), h|
+          h[format_message(k)] = format_message(v)
```

Comment:
This prints hashes in kind of a misleading way, because it will convert some non-string elements to strings.

```puppet
  $obj = {
    "foo" => 1,
    "bar" => 2
  }
  out::message($obj)
```

results in

```
{
  "foo": "1",
  "bar": "2"
}
```

Reply:
I would expect the array case to have similar behavior.